File: packages/games/pong/logic/src/state.ts
import {
  BALL_INITIAL_SPEED,
  COUNTDOWN_TICKS,
  FIELD_HEIGHT,
  FIELD_WIDTH,
} from './constants.ts';

/** Which end of the field a player defends. */
export type Side = 'left' | 'right';

/**
 * Where a match is in its life.
 *
 * `pointScored` is a phase of its own rather than a flag, because the pause
 * after a point is simulated like any other: it advances tick by tick, so the
 * server and the clients agree on exactly when play resumes.
 */
export type MatchPhase = 'waiting' | 'countdown' | 'playing' | 'pointScored' | 'finished';

export interface Paddle {
  /** Centre of the paddle along the y axis. */
  readonly y: number;
}

export interface Ball {
  readonly x: number;
  readonly y: number;
  readonly vx: number;
  readonly vy: number;
  /** Magnitude of the velocity, carried so a bounce can rebuild it exactly. */
  readonly speed: number;
}

export interface Score {
  readonly left: number;
  readonly right: number;
}

/** The complete simulation state. Everything else is derived from it. */
export interface PongState {
  readonly phase: MatchPhase;
  /** Ticks left in a timed phase. Zero outside `countdown` and `pointScored`. */
  readonly phaseTicks: number;
  readonly left: Paddle;
  readonly right: Paddle;
  readonly ball: Ball;
  readonly score: Score;
  /** Points completed so far, which chooses the serve angle. */
  readonly pointsPlayed: number;
  /** Side the next serve travels towards. */
  readonly serveTowards: Side;
  /** Winner, once the match is over. */
  readonly winner: Side | null;
}

/** What one player is pressing. Intent only, no position. */
export interface PaddleInput {
  readonly up: boolean;
  readonly down: boolean;
}

/** Both players' intent for a single tick. */
export interface PongInputs {
  readonly left: PaddleInput;
  readonly right: PaddleInput;
}

export const NO_INPUT: PaddleInput = { up: false, down: false };

/** A ball resting at the centre, as it sits between points. */
function restingBall(): Ball {
  return { x: FIELD_WIDTH / 2, y: FIELD_HEIGHT / 2, vx: 0, vy: 0, speed: BALL_INITIAL_SPEED };
}

/**
 * The state a match starts in, before anyone has joined.
 *
 * Callers move it to `countdown` with `startCountdown` once both seats are
 * filled.
 */
export function createInitialState(): PongState {
  return {
    phase: 'waiting',
    phaseTicks: 0,
    left: { y: FIELD_HEIGHT / 2 },
    right: { y: FIELD_HEIGHT / 2 },
    ball: restingBall(),
    score: { left: 0, right: 0 },
    pointsPlayed: 0,
    serveTowards: 'right',
    winner: null,
  };
}

/** Begins the three-second countdown that precedes the first rally. */
export function startCountdown(state: PongState): PongState {
  return { ...state, phase: 'countdown', phaseTicks: COUNTDOWN_TICKS, ball: restingBall() };
}

export { restingBall };

File: server/nakama/pong/state.go
package pong

// Phases a match moves through. The values match the TypeScript ones exactly,
// because both appear in the conformance vectors and on the wire.
const (
	PhaseWaiting     = "waiting"
	PhaseCountdown   = "countdown"
	PhasePlaying     = "playing"
	PhasePointScored = "pointScored"
	PhaseFinished    = "finished"
)

// Sides of the field.
const (
	SideLeft  = "left"
	SideRight = "right"
)

type Paddle struct {
	// Y is the centre of the paddle along the y axis.
	Y float64 `json:"y"`
}

type Ball struct {
	X  float64 `json:"x"`
	Y  float64 `json:"y"`
	VX float64 `json:"vx"`
	VY float64 `json:"vy"`
	// Speed is the magnitude of the velocity, carried so a bounce can rebuild
	// it exactly.
	Speed float64 `json:"speed"`
}

type Score struct {
	Left  int `json:"left"`
	Right int `json:"right"`
}

// PaddleInput is what one player is pressing. Intent only, no position.
type PaddleInput struct {
	Up   bool `json:"up"`
	Down bool `json:"down"`
}

// Inputs is both players' intent for a single tick.
type Inputs struct {
	Left  PaddleInput `json:"left"`
	Right PaddleInput `json:"right"`
}

// State is the complete simulation state.
//
// Every field is comparable, so two states can be checked for exact equality
// without a helper — which is what the conformance test relies on.
type State struct {
	Phase string `json:"phase"`
	// PhaseTicks is the number of ticks left in a timed phase, and zero
	// outside countdown and pointScored.
	PhaseTicks int    `json:"phaseTicks"`
	Left       Paddle `json:"left"`
	Right      Paddle `json:"right"`
	Ball       Ball   `json:"ball"`
	Score      Score  `json:"score"`
	// PointsPlayed chooses the serve angle.
	PointsPlayed int `json:"pointsPlayed"`
	// ServeTowards is the side the next serve travels towards.
	ServeTowards string `json:"serveTowards"`
	// Winner is empty until the match is over. JSON null decodes to empty.
	Winner string `json:"winner"`
}

func restingBall() Ball {
	return Ball{X: FieldWidth / 2, Y: FieldHeight / 2, VX: 0, VY: 0, Speed: BallInitialSpeed}
}

// NewState returns the state a match starts in, before anyone has joined.
func NewState() State {
	return State{
		Phase:        PhaseWaiting,
		PhaseTicks:   0,
		Left:         Paddle{Y: FieldHeight / 2},
		Right:        Paddle{Y: FieldHeight / 2},
		Ball:         restingBall(),
		Score:        Score{Left: 0, Right: 0},
		PointsPlayed: 0,
		ServeTowards: SideRight,
		Winner:       "",
	}
}

// StartCountdown begins the three-second countdown before the first rally.
func StartCountdown(state State) State {
	state.Phase = PhaseCountdown
	state.PhaseTicks = CountdownTicks
	state.Ball = restingBall()
	return state
}
